### PR TITLE
Helm ocp support

### LINKED ISSUE #60: Support helm chart on OCP
## Summary
Today our helm chart doesn't support OCP because we require the conjur image to run as root and therefore to enable anyuid permission to the SA that will run the conjur deployment.

## Expected Results
Install Conjur using helm chart on OCP 3.x 4.x

## Additional Information
Our new container-dap project already have the fixes required to run without root.
We can take the fixes from the container-dap project and port them to the OSS project.

### PR DESCRIPTION
**Pull request by guygiat**
### What does this PR do?
- Add support for helm chart installation on OCP

### What ticket does this PR close?
Resolves #60 

### What has been changed 
- Add Postgres OCP Service, StatefulSet, ConfigMap and SSL config file
- Edit deployment.yml to run conjur image on port 8080 (80 is prohibited on OCP)
- Edit deployment.yml to use container-dap Conjur and Nginx images for OCP
- Edit values.yml to include openshift section that can be run using `openshift.enabled=true `

#### Change log
- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [x] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

#### Not needed for this effort but related 
- [x] Push the new Conjur and Nginx images to the RedHat catalog for enabling operator deployment

#### TODO
- [x] Make a new Conjur and Nginx image based on the container-dap repo and publish it to CyberArk docker hub

### CODE REVIEW

**Comment by gmcbrien** [comment]
Thanks for the OCP Support. However, in our OCP 4.5.8, we're still having Postgres problems, with the 'classic' error:

```
mkdir: cannot create directory '/var/lib/pgsql/data/userdata': Permission denied
```

**Comment by guygiat** [comment]
That's weird, did you try with the following command?
helm install -n "$CONJUR_NAMESPACE" "$HELM_RELEASE" ./conjur-oss --set
openshift.enabled=true --set dataKey="$(docker run --rm cyberark/conjur
data-key generate)"

**Comment by gmcbrien** [comment]
The below email is classified: Internal

Hi Guy and thanks or getting back to me.

You may notice, I actually deleted my comment in your PR, just an hour or so ago. I found that your charts worked fine in a ‘playground’ Openshift cluster, so there is nothing wrong with the charts.

I have taken a deeper look and it appears it’s related to the creation of our PVs somehow. I managed to get it running this morning but am still not 100% sure what I’ve changed.

Anyway, thanks again for your useful helm charts and I’ll get back to you in case of further problems.

Grant

**Comment by gmcbrien** [comment]
The below email is classified: Internal

Hi again Guy,

Just curious – would you have any plans to upgrade the conjur-oss image in your charts to, say, 1.8.1?

Thanks a lot and regards,

Grant McBrien


From: Grant McBrien
Sent: 29 September 2020 14:08
To: 'cyberark/conjur-oss-helm-chart' <reply@reply.github.com>
Subject: RE: [cyberark/conjur-oss-helm-chart] Helm ocp support (#71)

The below email is classified: Internal

Hi Guy and thanks or getting back to me.

You may notice, I actually deleted my comment in your PR, just an hour or so ago. I found that your charts worked fine in a ‘playground’ Openshift cluster, so there is nothing wrong with the charts.

I have taken a deeper look and it appears it’s related to the creation of our PVs somehow. I managed to get it running this morning but am still not 100% sure what I’ve changed.

Anyway, thanks again for your useful helm charts and I’ll get back to you in case of further problems.

Grant

**Comment by guygiat** [comment]
Hi Grant,

We are in the making of a conjur image based on RedHat UBI which will be
available in the redhat registry as well as an Operator :)

**Comment by gmcbrien** [comment]
Hi,
Just redeployed and am getting an error which we didn't have before:

`"/usr/bin/exec-entrypoint: line 3: exec: server: not found"`

I see the image was updated 8 days ago. Would you have any ideas?

Thanks a lot!

**Comment by guygiat** [comment]
Hi Grant,

We did a couple of changes because it is still on undergoing work.
In the meantime you can try using guygiat/conjur-ubi-ocp for the
conjur image and guygiat/nginx-ubi for the nginx image.

Thanks.

**Review by diverdane — COMMENTED** [review]
@guygiat ,
Nice work, thank you so much for adding this support!

I've tested this on our OCP setup and it works (yay!), although this setup doesn't have a storage class, so I had to test without persistence.

My one overall comment is that I'd like to try to merge (or "interleave" is probably a better way to put it) the Kubernetes and OpenShift postgres deployments and values.yaml configs. It may help to use the same approaches for both as much as possible. For example, the OpenShift uses a ConfigMap for SSL settings, and Kubernetes uses command line arguments. It would be better to pick one of these approaches for both (if providing a command + args won't work for OpenShift, I can convert the Kubernetes deployment to use a ConfigMap).

I started to try combining these deployments myself... but I suspect that you probably started with this approach and then at some point things became too divergent? I'll continue to explore this, let me know if `there be dragons` with this.

**Review by diverdane — COMMENTED** [review]
Okay, I took a shot at merging/converging the postgres YAML files for OpenShift vs Kubernetes, and it turned out to have more differences than lines in common for OCP vs K8s (which you probably knew from your experience with this postgres image).

I'm wondering if a better approach would be to use the same image that you're using for OpenShift in the generic Kubernetes platforms. Obviously we'd want your changes to merge first, and tackle the Kubernetes changes separately.

File: conjur-oss/templates/postgres-ocp.yaml
Comment (guygiat) [reply conjur-oss/templates/postgres-ocp.yaml:84]: I copied the structure of `{{ .Release.Name }}-` from the current `deployment.yaml`, I have no problem to remove it though. WDYT?

File: conjur-oss/templates/postgres-ocp.yaml
Comment (guygiat) [reply conjur-oss/templates/postgres-ocp.yaml:100]: I copied the structure of `-ssl-` from the current deployment.yaml, I have no problem to remove it though. WDYT?

Comment (diverdane) [reply conjur-oss/templates/postgres-ocp.yaml:100]: I'd have a slight preference for dropping the `-ssl-`, just in case we add more Postgres config in the future, but your call.

File: conjur-oss/templates/postgres-ocp.yaml
Comment (diverdane) [reply conjur-oss/templates/postgres-ocp.yaml:84]: Ah... Yeah, I'd say leave it the way you have it to stay consistent. At some point we should consider simplifying the volume names, seems like unnecessary complexity.

**Review by diverdane — COMMENTED** [review]
I forgot to mention, any changes we make to the `values.yaml` will need corresponding
sanity checks added to the `values.schema.json`

File: conjur-oss/templates/postgres-ocp.yaml
Comment (guygiat) [reply conjur-oss/templates/postgres-ocp.yaml:84]: agree

File: conjur-oss/README.md
Comment (izgeri) [inline conjur-oss/README.md:52]: we are going to need to add a note here that this is community support for now, until our OCP clusters are able to support persistent volumes for automated testing.
```suggestion
- OCP 4.x+ (**Preliminary [Community level](https://github.com/cyberark/community/blob/master/Conjur/conventions/certification-levels.md#community) support) only**)
```

**Review by izgeri — CHANGES_REQUESTED** [review]
@guygiat this change is going to need an entry in CHANGELOG.md. I would suggest something like this under `Added`:
> The Conjur OSS helm chart has Community support for deploying Conjur OSS to OpenShift 4.x. [cyberark/conjur-oss-helm-chart#60](https://github.com/cyberark/conjur-oss-helm-chart/issues/60)

**Review by izgeri — APPROVED** [review]
The changes I requested are all set - @diverdane I leave the rest to your technical review.

**Review by diverdane — CHANGES_REQUESTED** [review]
Hi @guygiat ,
I'm trying to test these changes on a local Kubernetes cluster to make sure that we're not
causing any regressions, but I'm seeing some issues when I try to log into Conjur from a Conjur CLI pod:

```
root@conjur-cli-5bc479754b-tx92r:~# conjur authn login -u admin -p XXXXXX
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx</center>
</body>
</html>
error: 502 Bad Gateway
root@conjur-cli-5bc479754b-tx92r:~#
```
This could be operator error on my part, but can you hold off on merging this until I get a chance to figure out what's happening here?

**Comment by guygiat** [comment]
BTW, I don't have write access to this repo so you or Geri will have to merge it :)

**Comment by guygiat** [comment]
> Hi @guygiat ,
> I'm trying to test these changes on a local Kubernetes cluster to make sure that we're not
> causing any regressions, but I'm seeing some issues when I try to log into Conjur from a Conjur CLI pod:
> 
> ```
> root@conjur-cli-5bc479754b-tx92r:~# conjur authn login -u admin -p XXXXXX
> <html>
> <head><title>502 Bad Gateway</title></head>
> <body>
> <center><h1>502 Bad Gateway</h1></center>
> <hr><center>nginx</center>
> </body>
> </html>
> error: 502 Bad Gateway
> root@conjur-cli-5bc479754b-tx92r:~#
> ```
> 
> This could be operator error on my part, but can you hold off on merging this until I get a chance to figure out what's happening here?

fix in my last commit :)

**Review by diverdane — APPROVED** [review]
LGTM!

**Pull request merged**
